# E-Commerce Customer Analysis

In [1]:
import numpy as np
import pandas as pd

In [2]:
# understanding the quality of the dataset

#loading the dataset into a dataframe
df = pd.read_csv(r'\Users\oseni\Downloads\customer_sales - customer_sales.csv')

# displaying first five records

df.head()


,Customer ID,Customer Name,City,Age,Product Category,Sales,Order Status
0,CUST00001,Joseph Obi,Lagos,65,Beauty,61241.84,Completed
1,CUST00002,Tolu Onyeka,Abuja,55,Home Appliances,7969.88,Completed
2,CUST00003,Fatima Ogunleye,Ilorin,19,Accessories,49729.44,Pending
3,CUST00004,Mary Akinyemi,Ibadan,46,Office Supplies,69565.72,Returned
4,CUST00005,Kelechi Abubakar,Owerri,45,Sports,69485.89,Completed


In [3]:
# Number of rows and columns

df.shape

(5000, 7)

In [4]:
# check column names

df.columns

Index(['Customer ID', 'Customer Name', 'City', 'Age', 'Product Category',
       'Sales', 'Order Status'],
      dtype='str')

In [5]:
# check data type

df.dtypes

Customer ID             str
Customer Name           str
City                    str
Age                   int64
Product Category        str
Sales               float64
Order Status            str
dtype: object

In [6]:
# Statistics

df.describe()

,Age,Sales
count,5000.000000,5000.000000
mean,43.905400,124393.017430
std,15.358881,71795.872349
min,18.000000,37.880000
25%,31.000000,61591.572500
50%,44.000000,124713.010000
75%,57.000000,186804.632500
max,70.000000,249810.490000


In [7]:
# check the statistics

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Customer ID       5000 non-null   str    
 1   Customer Name     5000 non-null   str    
 2   City              5000 non-null   str    
 3   Age               5000 non-null   int64  
 4   Product Category  5000 non-null   str    
 5   Sales             5000 non-null   float64
 6   Order Status      5000 non-null   str    
dtypes: float64(1), int64(1), str(5)
memory usage: 273.6 KB


In [8]:
# check for the number of missing city values

empty_city = df["City"].isna().sum()
print("Empty Cell:", empty_city)

Empty Cell: 0


In [9]:
# Text values appearing where numerical values are expected

empty_age = pd.to_numeric(df['Age'], errors = 'coerce').isna().sum()
print("Empty Age:", empty_age)


Empty Age: 0


In [10]:
## Check for duplicates

duplicate_cell = df.duplicated().sum()
print("Duplicate Cell:", duplicate_cell)


Duplicate Cell: 0


In [11]:
# customers in each city

df['City'].value_counts()

City
Enugu            364
Abuja            360
Calabar          358
Warri            348
Uyo              346
Kaduna           338
Kano             337
Port Harcourt    336
Ilorin           331
Abeokuta         327
Benin City       325
Jos              321
Lagos            308
Owerri           303
Ibadan           298
Name: count, dtype: int64

In [12]:
# Average sales value

mean = df['Sales'].mean().round(4)
print(f"Average Value: {mean}")

Average Value: 124393.0174


In [13]:
df['Product Category'].value_counts()

Product Category
Home Appliances    549
Office Supplies    524
Accessories        507
Books              505
Beauty             499
Clothing           497
Sports             484
Electronics        483
Furniture          478
Groceries          474
Name: count, dtype: int64

In [14]:
# Frequency of products

frequent_product = df['Product Category'].value_counts().idxmax()
print(f"Most Frequent Product: {frequent_product}")


Most Frequent Product: Home Appliances


In [15]:
# Completed orders

completed_orders = (df['Order Status'] == "Completed").sum()
print(f"Number of Completed Orders: {completed_orders}")


Number of Completed Orders: 3099


In [16]:
# check for empty columns

df.isna().sum()

Customer ID         0
Customer Name       0
City                0
Age                 0
Product Category    0
Sales               0
Order Status        0
dtype: int64

In [17]:
# Create a condition based column

df['Customer Value'] = np.where(df['Sales'] >= 100000, "High value", "Standard")
df

,Customer ID,Customer Name,City,Age,Product Category,Sales,Order Status,Customer Value
0,CUST00001,Joseph Obi,Lagos,65,Beauty,61241.84,Completed,Standard
1,CUST00002,Tolu Onyeka,Abuja,55,Home Appliances,7969.88,Completed,Standard
2,CUST00003,Fatima Ogunleye,Ilorin,19,Accessories,49729.44,Pending,Standard
3,CUST00004,Mary Akinyemi,Ibadan,46,Office Supplies,69565.72,Returned,Standard
4,CUST00005,Kelechi Abubakar,Owerri,45,Sports,69485.89,Completed,Standard
...,...,...,...,...,...,...,...,...
4995,CUST04996,Hauwa Onyeka,Calabar,69,Furniture,50435.76,Completed,Standard
4996,CUST04997,Esther Ogunleye,Kaduna,27,Furniture,172206.40,Completed,High value
4997,CUST04998,Mary Okafor,Enugu,21,Clothing,216110.54,Completed,High value
4998,CUST04999,Joy Balogun,Ibadan,26,Home Appliances,121369.50,Completed,High value


In [18]:
# Determine customers with high values

high_value_customers = df['Customer Value'].value_counts().max()
print(f"Number of high Value Customers is: {high_value_customers}")


Number of high Value Customers is: 2983


# Hospital Patient Data Cleaning

In [19]:
# loading the dataset into a dataframe

df_1 = pd.read_csv(r'\Users\oseni\Downloads\hospital_patient_data - hospital_patient_data_cleaning_practice.csv')

In [20]:
# exploration of data

df_1.head()


,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
0,1745,Sani Olawale,58.0,Male,Asaba,3,2026,Specialist Consultation
1,1679,Daniel James,49.0,Male,Owerri,10,2025,Specialist Consultation
2,3291,John Ibekwe,20.0,Female,Asaba,9,2026,Routine Check
3,3489,Bola Sani,56.0,Female,Ilorin,2,2024,Routine Check
4,1943,Aisha Chukwu,50.0,Male,Lagos,11,2026,Routine Check


In [21]:
df_1.shape

(3000, 8)

In [22]:
df_1.info()

<class 'pandas.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Patient_ID    3000 non-null   int64  
 1   Patient_Name  3000 non-null   str    
 2   Age           2955 non-null   float64
 3   Gender        3000 non-null   str    
 4   City          2881 non-null   str    
 5   Visit_Count   2959 non-null   str    
 6   Record_Year   3000 non-null   int64  
 7   Visit_Type    3000 non-null   str    
dtypes: float64(1), int64(2), str(5)
memory usage: 187.6 KB


In [23]:
df_1.describe()

,Patient_ID,Age,Record_Year
count,3000.000000,2955.000000,3000.000000
mean,2511.946667,49.008460,2025.239000
std,905.746037,18.137469,0.763811
min,1001.000000,0.000000,2024.000000
25%,1751.000000,34.000000,2025.000000
50%,2504.500000,49.000000,2025.000000
75%,3255.250000,65.000000,2026.000000
max,9005.000000,99.000000,2026.000000


In [24]:
## Records with missing Information

df_1.isna().sum()

Patient_ID        0
Patient_Name      0
Age              45
Gender            0
City            119
Visit_Count      41
Record_Year       0
Visit_Type        0
dtype: int64

#### Discovery
##### There are 3000 rows and 8 columns, The data type of age is float instead of integer and it contains about 45 null cells
##### It was also discoved that city, and visit_count also contain some empty values and visit count has a string data type

In [25]:
# check for duplicate data

duplicate_data = df_1.duplicated().sum()
print("Duplicate Data:", duplicate_data)

Duplicate Data: 4


In [26]:
# drop duplicate data

df_1 = df_1.drop_duplicates()

In [27]:
# check if patient id has duplicates

df_1.duplicated(subset = ['Patient_ID']).sum()

np.int64(0)

In [28]:
# observe columns using unique 

df_1['Gender'].unique()

<StringArray>
['Male', 'Female', 'Unknown']
Length: 3, dtype: str

In [29]:
df_1['Gender'].value_counts()

Gender
Female     1504
Male       1487
Unknown       5
Name: count, dtype: int64

In [30]:
df_1['Record_Year'].unique()

array([2026, 2025, 2024])

In [31]:
df_1['Visit_Type'].unique()

<StringArray>
['Specialist Consultation', 'Routine Check', 'Follow-up', 'Chronic Care',
 'Test']
Length: 5, dtype: str

In [32]:
df_1['City'].value_counts()

City
Calabar          177
Uyo              173
Warri            170
Zaria            170
Enugu            165
Kano             165
Asaba            164
Ilorin           161
Kaduna           160
Benin City       158
Abeokuta         154
Akure            150
Port Harcourt    149
Lagos            147
Abuja            143
Jos              140
Owerri           139
Ibadan           128
lagos             15
ABUJA             15
ibadan            15
Port-Harcourt     15
TEST               5
Name: count, dtype: int64

In [33]:
df_1['City'].unique()

<StringArray>
[        'Asaba',        'Owerri',        'Ilorin',         'Lagos',
         'Enugu',        'Kaduna',         'Warri',      'Abeokuta',
       'Calabar',          'Kano',        'Ibadan',             nan,
          'TEST',           'Jos',         'Akure',           'Uyo',
         'Zaria',    'Benin City',         'Abuja', 'Port Harcourt',
         'lagos',         'ABUJA',        'ibadan', 'Port-Harcourt']
Length: 24, dtype: str

In [34]:
# remove trailing spaces and change data format
df_1['City'] = df_1['City'].astype(str).str.strip().str.title()

# replace values with the correct format
df_1['City'] = df_1['City'].str.replace("Port-Harcourt", "Port Harcourt")

#remove wrong values
df_1 = df_1[df_1['City'] != 'Test']


In [35]:
df_1['City'].unique()

<StringArray>
[        'Asaba',        'Owerri',        'Ilorin',         'Lagos',
         'Enugu',        'Kaduna',         'Warri',      'Abeokuta',
       'Calabar',          'Kano',        'Ibadan',             nan,
           'Jos',         'Akure',           'Uyo',         'Zaria',
    'Benin City',         'Abuja', 'Port Harcourt']
Length: 19, dtype: str

In [36]:
# fill up empty spaces
df_1['City'] = df_1['City'].str.strip().fillna("Unknown")

In [37]:
df_1.info()

<class 'pandas.DataFrame'>
Index: 2991 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Patient_ID    2991 non-null   int64  
 1   Patient_Name  2991 non-null   str    
 2   Age           2946 non-null   float64
 3   Gender        2991 non-null   str    
 4   City          2991 non-null   str    
 5   Visit_Count   2953 non-null   str    
 6   Record_Year   2991 non-null   int64  
 7   Visit_Type    2991 non-null   str    
dtypes: float64(1), int64(2), str(5)
memory usage: 210.3 KB


In [38]:
# Clean age data

df_1['Age'] = pd.to_numeric(df_1['Age'], errors = 'coerce')
median_age = df_1['Age'].median()

df_1['Age'] = df_1['Age'].fillna(median_age).astype(int)

df_1.head(15)

,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
0,1745,Sani Olawale,58,Male,Asaba,3,2026,Specialist Consultation
1,1679,Daniel James,49,Male,Owerri,10,2025,Specialist Consultation
2,3291,John Ibekwe,20,Female,Asaba,9,2026,Routine Check
3,3489,Bola Sani,56,Female,Ilorin,2,2024,Routine Check
4,1943,Aisha Chukwu,50,Male,Lagos,11,2026,Routine Check
5,3956,Yusuf Ibrahim,43,Male,Enugu,7,2025,Follow-up
6,1290,Sarah Olawale,25,Male,Kaduna,six,2026,Follow-up
7,1434,Precious Okafor,37,Male,Warri,1,2025,Specialist Consultation
8,3830,Femi Garba,60,Female,Abeokuta,6,2026,Chronic Care
9,3475,Zainab Peter,21,Male,Calabar,5,2026,Follow-up


In [39]:
df_1.info()

<class 'pandas.DataFrame'>
Index: 2991 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Patient_ID    2991 non-null   int64
 1   Patient_Name  2991 non-null   str  
 2   Age           2991 non-null   int64
 3   Gender        2991 non-null   str  
 4   City          2991 non-null   str  
 5   Visit_Count   2953 non-null   str  
 6   Record_Year   2991 non-null   int64
 7   Visit_Type    2991 non-null   str  
dtypes: int64(3), str(5)
memory usage: 210.3 KB


In [40]:
df_1['Visit_Count'].unique()

<StringArray>
[       '3',       '10',        '9',        '2',       '11',        '7',
      'six',        '1',        '6',        '5',        nan,       '12',
        '4',        '8', '4 visits',    'seven',      'ten']
Length: 17, dtype: str

In [41]:
# clean visit count and convert to proper format

df_1['Visit_Count'] = df_1['Visit_Count'].astype(str).str.replace('visits', '', case = False).str.strip()

word_to_num = {'six': '6',
              'seven': '7',
              'ten': '10',
              'nan': 'np.nan'}
df_1['Visit_Count'] = df_1['Visit_Count'].str.lower().replace(word_to_num)


df_1['Visit_Count'] = pd.to_numeric(df_1['Visit_Count'], errors = 'coerce')

df_1['Visit_Count'] = df_1['Visit_Count'].fillna(0).astype(int)

df_1.info()


<class 'pandas.DataFrame'>
Index: 2991 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Patient_ID    2991 non-null   int64
 1   Patient_Name  2991 non-null   str  
 2   Age           2991 non-null   int64
 3   Gender        2991 non-null   str  
 4   City          2991 non-null   str  
 5   Visit_Count   2991 non-null   int64
 6   Record_Year   2991 non-null   int64
 7   Visit_Type    2991 non-null   str  
dtypes: int64(4), str(4)
memory usage: 210.3 KB


In [42]:
df_1.head(15)

,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
0,1745,Sani Olawale,58,Male,Asaba,3,2026,Specialist Consultation
1,1679,Daniel James,49,Male,Owerri,10,2025,Specialist Consultation
2,3291,John Ibekwe,20,Female,Asaba,9,2026,Routine Check
3,3489,Bola Sani,56,Female,Ilorin,2,2024,Routine Check
4,1943,Aisha Chukwu,50,Male,Lagos,11,2026,Routine Check
5,3956,Yusuf Ibrahim,43,Male,Enugu,7,2025,Follow-up
6,1290,Sarah Olawale,25,Male,Kaduna,6,2026,Follow-up
7,1434,Precious Okafor,37,Male,Warri,1,2025,Specialist Consultation
8,3830,Femi Garba,60,Female,Abeokuta,6,2026,Chronic Care
9,3475,Zainab Peter,21,Male,Calabar,5,2026,Follow-up


In [43]:
# create new column with a conditional based format

df_1['Patient_Category'] = np.where(df_1['Visit_Count'] >= 5, "Frequent Patient" , "Occasional Patient")
df_1.head(15)

,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type,Patient_Category
0,1745,Sani Olawale,58,Male,Asaba,3,2026,Specialist Consultation,Occasional Patient
1,1679,Daniel James,49,Male,Owerri,10,2025,Specialist Consultation,Frequent Patient
2,3291,John Ibekwe,20,Female,Asaba,9,2026,Routine Check,Frequent Patient
3,3489,Bola Sani,56,Female,Ilorin,2,2024,Routine Check,Occasional Patient
4,1943,Aisha Chukwu,50,Male,Lagos,11,2026,Routine Check,Frequent Patient
5,3956,Yusuf Ibrahim,43,Male,Enugu,7,2025,Follow-up,Frequent Patient
6,1290,Sarah Olawale,25,Male,Kaduna,6,2026,Follow-up,Frequent Patient
7,1434,Precious Okafor,37,Male,Warri,1,2025,Specialist Consultation,Occasional Patient
8,3830,Femi Garba,60,Female,Abeokuta,6,2026,Chronic Care,Frequent Patient
9,3475,Zainab Peter,21,Male,Calabar,5,2026,Follow-up,Frequent Patient


In [44]:
## Number of patients by City

df_1['City'].value_counts()


City
Calabar          177
Uyo              173
Warri            170
Zaria            170
Enugu            165
Kano             165
Asaba            164
Port Harcourt    164
Lagos            162
Ilorin           161
Kaduna           160
Benin City       158
Abuja            158
Abeokuta         154
Akure            150
Ibadan           143
Jos              140
Owerri           139
Unknown          118
Name: count, dtype: int64

In [45]:
## Number of Frequent Patients

frequent_patient = df_1['Patient_Category'].value_counts().max()
print("Number of Frequent Patient:", frequent_patient)

Number of Frequent Patient: 1986


In [46]:
## Average patient visit count

average_patient_visit = df_1['Visit_Count'].mean()
print("Average Visit Count:", average_patient_visit.round(3))

Average Visit Count: 6.441


In [47]:
## Number of Records remaining after cleaning

df_1.info()

<class 'pandas.DataFrame'>
Index: 2991 entries, 0 to 2999
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   Patient_ID        2991 non-null   int64
 1   Patient_Name      2991 non-null   str  
 2   Age               2991 non-null   int64
 3   Gender            2991 non-null   str  
 4   City              2991 non-null   str  
 5   Visit_Count       2991 non-null   int64
 6   Record_Year       2991 non-null   int64
 7   Visit_Type        2991 non-null   str  
 8   Patient_Category  2991 non-null   str  
dtypes: int64(4), str(5)
memory usage: 233.7 KB


In [48]:
# Length of dataset after cleaning

Length = len(df_1)
print("Dataset after Cleaning:", Length)

Dataset after Cleaning: 2991


In [49]:
# create a summary table

number_of_patients = df_1['Patient_Category'].value_counts()
average_visit = df_1.groupby('Patient_Category')['Visit_Count'].mean()

pd.concat([number_of_patients,average_visit.round(3)], axis = 1, keys = ['Number of Patient', 'Average Visit'])

,Number of Patient,Average Visit
Patient_Category,,
Frequent Patient,1986,8.477
Occasional Patient,1005,2.419


#### Observation
##### The table shows that the hospital has 1986 Frequent patient who visit on an average of 8.5 times, compared to 1005 Occasiona Patients who average 2.4 visits. 
##### This implies that the Frequent patient drive the vast majority of hospital daily traffic and workload despite making up to  two third of the headcounts. Consequently, hospitals need to create structures ans scheduling programmes to prevent overcrowding.

# Banking Customer Transaction Analysis

In [50]:
# loading the dataset into a dataframe

df_2 = pd.read_csv(r'\Users\oseni\Downloads\Banking_Customer_Transation - Banking_Customer_Transation.csv')


In [51]:
# data exploration

df_2.shape

(3000, 6)

In [52]:
df_2.columns

Index(['Customer ID', 'Customer Name', 'City', 'Transaction Type', 'Amount',
       'Account Status'],
      dtype='str')

In [53]:
df_2.head(10)

,Customer ID,Customer Name,City,Transaction Type,Amount,Account Status
0,CUST00585,Tunde Okafor,Kaduna,Online Transfer,37974.75,Active
1,CUST01061,Bola Uche,Lagos,Deposit,9896.89,Active
2,CUST00077,Kelechi Nwosu,Abuja,Transfer,59302.75,Active
3,CUST00898,Tunde Abubakar,Benin City,Deposit,71719.97,Dormant
4,CUST00594,Amina Ojo,Lagos,Online Transfer,3979.55,Active
5,CUST00796,Amaka Adebayo,Benin City,Online Transfer,7614.28,Active
6,CUST00107,Ada Bello,Lagos,Deposit,31818.41,Active
7,CUST00957,Zainab Uche,Port Harcourt,Deposit,20408.71,Active
8,CUST00013,Obinna Ekanem,Lagos,Transfer,27533.5,Active
9,CUST00998,Ada Abubakar,Ibadan,Withdrawal,11931.24,Active


In [54]:
df_2.describe()

,Customer ID,Customer Name,City,Transaction Type,Amount,Account Status
count,3000,2990,2976,2982,2884,2988
unique,1031,400,10,7,2803,4
top,CUST00821,Femi Ekanem,Lagos,Transfer,Not Available,Active
freq,9,16,731,548,20,2371


In [55]:
df_2.info()

<class 'pandas.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   Customer ID       3000 non-null   str  
 1   Customer Name     2990 non-null   str  
 2   City              2976 non-null   str  
 3   Transaction Type  2982 non-null   str  
 4   Amount            2884 non-null   str  
 5   Account Status    2988 non-null   str  
dtypes: str(6)
memory usage: 140.8 KB


In [56]:
# check for empty cells

df_2.isna().sum()

Customer ID           0
Customer Name        10
City                 24
Transaction Type     18
Amount              116
Account Status       12
dtype: int64

In [57]:
# check for duplicates

duplicate_data = df_2.duplicated().sum()
print(duplicate_data)

0


In [58]:
## Check if Customername are truly empty

df_2['Customer Name'] = df_2['Customer Name'].astype(str).str.strip()
df_2['Customer Name'] = df_2['Customer Name'].replace('', np.nan)
df_2['Customer Name'].isna().sum()

np.int64(10)

In [59]:
# Check if city is truly empty

df_2['City'] = df_2['City'].astype(str).str.strip()
df_2['City'] = df_2['City'].replace('', np.nan)
df_2['City'].isna().sum()

np.int64(24)

In [60]:
# Check if Transacction type is truly empty

df_2['Transaction Type'] = df_2['Transaction Type'].astype(str).str.strip().replace('', np.nan)
df_2['Transaction Type'].isna().sum()

np.int64(18)

In [61]:
## Check if Account Status is truly empty

df_2['Account Status'] = df_2['Account Status'].astype(str).str.strip().replace('', np.nan)
df_2['Account Status'].isna().sum()

np.int64(12)

In [62]:
# Check for spaces that are not empty and cant be coerced to numbers

df_2['Amount'] = df_2['Amount'].astype(str).str.strip().replace('', np.nan)

text_to_number = pd.to_numeric(df_2['Amount'], errors= 'coerce')

failed_row = df_2[df_2['Amount'].notna() & text_to_number.isna()]

len(failed_row)

80

In [63]:
failed_row['Amount'].unique()

<StringArray>
['Error', 'Not Available', 'Pending', 'Unknown', '##VALUE!', '1O000',
 '₦12,500']
Length: 7, dtype: str

In [64]:
# clean identified error

df_2['Amount'] = df_2['Amount'].str.replace('₦', '', regex = False)
df_2['Amount'] = df_2['Amount'].str.replace(',', '', regex = False)
df_2['Amount'] = df_2['Amount'].str.replace('O', '0', regex = False)

df_2['Amount'] = pd.to_numeric(df_2['Amount'], errors = 'coerce')
df_2['Amount'].isna().sum()

np.int64(186)

In [65]:
# fill up empty cells

df_2['Customer Name'] = df_2['Customer Name'].fillna('Unknown')
df_2['City'] = df_2['City'].fillna('Unknown')
df_2['Transaction Type'] = df_2['Transaction Type'].fillna('Unknown')
df_2['Account Status'] = df_2['Account Status'].fillna('Unknown')

In [66]:
df_2.info()

<class 'pandas.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Customer ID       3000 non-null   str    
 1   Customer Name     3000 non-null   str    
 2   City              3000 non-null   str    
 3   Transaction Type  3000 non-null   str    
 4   Amount            2814 non-null   float64
 5   Account Status    3000 non-null   str    
dtypes: float64(1), str(5)
memory usage: 140.8 KB


In [67]:
# create new column using conditional based format

df_2['Transaction_Category'] = np.where(df_2['Amount'] >= 100000, "Large Transaction", "Regular Transaction")
df_2

,Customer ID,Customer Name,City,Transaction Type,Amount,Account Status,Transaction_Category
0,CUST00585,Tunde Okafor,Kaduna,Online Transfer,37974.75,Active,Regular Transaction
1,CUST01061,Bola Uche,Lagos,Deposit,9896.89,Active,Regular Transaction
2,CUST00077,Kelechi Nwosu,Abuja,Transfer,59302.75,Active,Regular Transaction
3,CUST00898,Tunde Abubakar,Benin City,Deposit,71719.97,Dormant,Regular Transaction
4,CUST00594,Amina Ojo,Lagos,Online Transfer,3979.55,Active,Regular Transaction
...,...,...,...,...,...,...,...
2995,CUST00610,Ifeoma Nnamdi,Lagos,Withdrawal,13010.32,Active,Regular Transaction
2996,CUST01014,Bola Nwosu,Ibadan,POS Purchase,8349.08,Active,Regular Transaction
2997,CUST00634,Zainab Eze,Kaduna,Withdrawal,288077.40,Active,Large Transaction
2998,CUST00552,Ngozi Ibrahim,Jos,Online Transfer,10000.00,Dormant,Regular Transaction


In [68]:
# Transactions in each city

df_2['City'].value_counts()

City
Lagos            731
Abuja            495
Port Harcourt    353
Ibadan           314
Kano             302
Enugu            193
Kaduna           184
Benin City       165
Aba              130
Jos              109
Unknown           24
Name: count, dtype: int64

In [69]:
# Average Transaction amount

transaction_amount = df_2['Amount'].mean()
print("Average Transaction Amount:", transaction_amount.round(3))

Average Transaction Amount: 45076.655


In [70]:
# Transaction Category that are Large transactions

large_transactions = df_2['Transaction_Category'].value_counts()
large_transactions 

Transaction_Category
Regular Transaction    2724
Large Transaction       276
Name: count, dtype: int64

In [71]:
# Value of Large transaction

print("Large Transaction:", large_transactions.min())

Large Transaction: 276


In [72]:
# Transaction with the most frequen transaction type

frequent_transaction_type = df_2['Transaction Type'].value_counts().idxmax()
print("Most Frequent Transaction Type:", frequent_transaction_type)

Most Frequent Transaction Type: Transfer


In [73]:
# Average transaction amount by transaction type

df_2.groupby('Transaction Type')['Amount'].mean().round(3)

Transaction Type
ATM Withdrawal     36960.128
Deposit            44458.180
Online Transfer    46682.385
POS Purchase       50549.251
Payment            40820.957
Transfer           46319.721
Unknown            46222.925
Withdrawal         48961.159
Name: Amount, dtype: float64

In [74]:
# Average transaction amount by City

df_2.groupby('City')['Amount'].mean().round(3)

City
Aba              50332.422
Abuja            46284.952
Benin City       52154.374
Enugu            40047.125
Ibadan           42323.278
Jos              44224.250
Kaduna           46313.136
Kano             43659.664
Lagos            46905.361
Port Harcourt    41283.748
Unknown          33783.731
Name: Amount, dtype: float64

In [75]:
# city with the highest average transaction amount

highest_average_city = df_2.groupby('City')['Amount'].mean().round(3)
print("City with the highest average transaction amount:", highest_average_city.idxmax())


City with the highest average transaction amount: Benin City


##### Findings
##### This showed that Benin had the highest average transaction amount meaning that, they spend more money per transaction than any other city making it the most valuable region to generate high revenue. 
##### The business could decide to allocate  budget to improve the inventory level to maximize returns

# Energy Consumption Analysis

In [76]:
# loading the dataset into a dataframe

df_3 = pd.read_csv(r'\Users\oseni\Downloads\Energy_Consumption - Energy_Consumption.csv')

In [ ]:
# data exploration

df_3.info()

<class 'pandas.DataFrame'>
RangeIndex: 3012 entries, 0 to 3011
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   Customer ID          3012 non-null   str  
 1   City                 2970 non-null   str  
 2   Customer Type        2994 non-null   str  
 3   Monthly Consumption  2894 non-null   str  
 4   Account Status       2997 non-null   str  
 5   Energy Source        2992 non-null   str  
dtypes: str(6)
memory usage: 141.3 KB


In [ ]:
df_3.shape

In [ ]:
df_3.describe()

In [ ]:
df_3.head(10)

In [ ]:
# checking for empty cells

df_3.isna().sum()

In [ ]:
# checking for duplicates

duplicate_data = df_3.duplicated().sum()
print("Duplicates:", duplicate_data)

In [ ]:
# removing duplicates

df_3 = df_3.drop_duplicates()

In [ ]:
df_3.info()

In [ ]:
# check for hidden texts or values

df_3['City'].unique()

In [ ]:
# check if the empty cells are actually empty

df_3['City'] = df_3['City'].astype(str).str.strip()
df_3['City'] = df_3['City'].replace('', np.nan)
df_3['City'].isna().sum()

In [ ]:
df_3.info()

In [ ]:
# check if customer type is actually empty

df_3['Customer Type'] = df_3['Customer Type'].astype(str).str.strip()
df_3['Customer Type'] = df_3['Customer Type'].replace('', np.nan)
df_3['Customer Type'].isna().sum()

In [ ]:
# check for rows that failed to format to numbers

df_3['Monthly Consumption'] = df_3['Monthly Consumption'].astype(str).str.strip().replace('', np.nan)

text_to_number = pd.to_numeric(df_3['Monthly Consumption'], errors= 'coerce')

failed_row = df_3[df_3['Monthly Consumption'].notna() & text_to_number.isna()]

len(failed_row)

In [ ]:
failed_row['Monthly Consumption'].unique()

In [ ]:
# Cleaning the textual symbols

df_3['Monthly Consumption'] = df_3['Monthly Consumption'].str.replace(',', '', regex = False)
df_3['Monthly Consumption'] = df_3['Monthly Consumption'].str.replace('O', '0', regex = False)

df_3['Monthly Consumption'] = pd.to_numeric(df_3['Monthly Consumption'], errors = 'coerce')
df_3['Monthly Consumption'].isna().sum()

In [ ]:
# check if Account Status is actually empty

df_3['Account Status'] = df_3['Account Status'].astype(str).str.strip()
df_3['Account Status'] = df_3['Account Status'].replace('', np.nan)
df_3['Account Status'].isna().sum()

In [ ]:
# check if Energy Source is actually empty

df_3['Energy Source'] = df_3['Energy Source'].astype(str).str.strip()
df_3['Energy Source'] = df_3['Energy Source'].replace('', np.nan)
df_3['Energy Source'].isna().sum()

In [ ]:
# Check if the are all empty at some point

check_blank_rows = ['City',	'Customer Type', 'Monthly Consumption',	'Account Status', 'Energy Source']
df_3 = df_3.dropna(subset = check_blank_rows, how ='all')
len(df_3)

In [ ]:
# Fill up dataset

df_3['City'] = df_3['City'].fillna('Not Provided')
df_3['Customer Type'] = df_3['Customer Type'].fillna('Unknown')
df_3['Account Status'] = df_3['Account Status'].fillna('Unknown')
df_3['Energy Source'] = df_3['Energy Source'].fillna('Not provided')

In [ ]:
df_3.info()

In [ ]:
# Create consumption category

df_3['Consumption_Category'] = np.where(df_3['Monthly Consumption'] >= 500, "High Consumption", "Low Consumption")
df_3

In [ ]:
# Number of customers in each city.

df_3['City'].value_counts()


In [ ]:
# Average consumption.

average_consumption = df_3['Monthly Consumption'].mean()
print("Average Consumption:", average_consumption.round(3) )



In [ ]:
# Number of high-consumption customers.

high_consumption_users = df_3['Consumption_Category'].value_counts().min()
print("Number of High consumption Users:", high_consumption_users )

In [ ]:
# Average consumption by customer type.

average_consumption_customertype = df_3.groupby('Customer Type')['Monthly Consumption'].mean()
print(average_consumption_customertype.round(3))

In [ ]:
# Number of customers using each energy source

df_3['Energy Source'].value_counts()

In [ ]:
# creating summary table

customer_type = df_3['Customer Type'].value_counts()
average_Consumption = df_3.groupby('Customer Type')['Monthly Consumption'].mean()

pd.concat([customer_type, average_Consumption.round(3)], axis = 1, keys = ['Number of Customers', 'Average Consumption'])

##### Findings

##### Industrial customers represernts the smallest group yet drive the highest average consumption by 6464.326kwh per month
##### Residential users make up the vast majority of customer base yet have the lowest individual elctricity consumption footprint
##### The electricity company should redefine industrial operational factors to redefine high demand grid forecasts
